bpo-31182: enhancements to tarfile and zipfile's CLI

GadgetSteve opened the pull request:
Added command line -m functionality to Lib\tarfile.py and Lib\zipile.py as follows:

 - Wildcards (glob style)
 - Recursion
 - Add to existing repository, (uncompressed only for tars).

<!-- issue-number: bpo-31182 -->
https://bugs.python.org/issue31182
<!-- /issue-number -->

GadgetSteve committed:
bpo-300156: Add compressions to command line helps for -m tarfile -h

GadgetSteve committed:
bpo-300156: Lib/tarfile.py Move compressions to epilog & added info on recurse of directories. Added --add option.

GadgetSteve committed:
Rationalised tarfile

GadgetSteve committed:
Rationalised zipfile and added epilog

GadgetSteve committed:
Rationalised list option by adding verbose to printdir

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
Added glob wildcard matching and recursion to Lib\zipfile.py

GadgetSteve committed:
Added glob wildcard matching and recursion to Lib\tarfile.py

GadgetSteve committed:
Tweek zipfile so that the default list verbosity is as expected and modified test_zipfile to test all verbosity levels of list command

GadgetSteve committed:
Address travis test failures in tar_file.

GadgetSteve committed:
Added test of zipfile command --append

GadgetSteve committed:
Added basic wildcard test to test_zipfile

GadgetSteve committed:
A little more info on zipfile -v -c and added test for **\* recursion

GadgetSteve committed:
Added news message

GadgetSteve committed:
Added diagnostic message to test that passes localy and fails on travis.

GadgetSteve committed:
Changed test limits on failing recursive test, (seems to be platform depenedent).

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
Added skip attempt to zip file into itself, (especially likely with wildcards).

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
Skip add modified to use absolute filenames.

GadgetSteve committed:
Address travis fail where file handles are passed instead of names.

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
2nd try to address travis fail where file handles are passed instead of names.

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2447,20 +2447,50 @@ def is_tarfile(name):
     description = 'A simple command-line interface for tarfile module.'
-    parser = argparse.ArgumentParser(description=description)
+    epilog= ' '.join([
+        'When creating compression is determined by a tarfile',
```
"When creating a tarfile, compression is determined by the filename's final extension of any of %s. "

GadgetSteve replied:
Good point, much clearer.

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2447,20 +2447,50 @@ def is_tarfile(name):
+                       help='Create tarfile from files.')
+    group.add_argument('-a', '--add', nargs='+',
+                       metavar=('<name>', '<file|dir>'),
+                       help='Append files to a tarfile, (.tar only).')
```
Append files to a tarfile (uncompressed only)?

GadgetSteve replied:
Yep, clearer.

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2509,28 +2539,53 @@ def main():
         tar_mode = 'w:' + compressions[ext] if ext in compressions else 'w'
         tar_files = args.create
+        if args.verbose:
+            print('Create {!r}:'.format(tar_name))
```
"Creating {!r}".format(tar_name)

GadgetSteve replied:
Done.

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2509,28 +2539,53 @@ def main():
+        # Actual adding files is done later.
+    elif args.add is not None:
+        tar_name = args.add.pop(0)
+        _, ext = os.path.splitext(tar_name)
```
Might I suggest a bit better name than '_'? If it's unused, which it appears to be, ignored_name would do fine.

GadgetSteve replied:
Used `dummy` as ignored by pylint.

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2509,28 +2539,53 @@ def main():
+        tar_files = args.add
         if args.verbose:
-            print('{!r} file created.'.format(tar_name))
+            print('Add to {!r}:'.format(tar_name))
```
"Adding to {!r}:".format(tar_name)

GadgetSteve replied:
Changed to adding if the file exists and to creating if it doesn't

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2509,28 +2539,53 @@ def main():
+        if ext in compressions:
+            parser.exit(
+                1,
+                "Error: Add is not a supported operation for compressed tar files")
```
Add or append? And you might give a suggestion "; uncompress first".

GadgetSteve replied:
There currently isn't an option to unpack to tar file and in many cases unpacking would probably defeat the point of an append. I could possibly add an uncompress option to extract the tar file from the compressed formats then this might be an option.
Removed error if file does not exist, as it is then a create, and added the need to unpack and create.

drallensmith commented on Lib/tarfile.py:
```diff
@@ -2509,28 +2539,53 @@ def main():
+                    else:
+                        # Not a file or dir so possible wildcard
+                        if args.verbose:
+                            print("Possible  Wildcard Add!", file_name)
```
print("Possible wildcard add {!r}".format(file_name)) - will make sure it's quoted properly
And, BTW, do you want the prints for verbose going to stdout (as they appear to be) or stderr?

GadgetSteve replied:
Good point I have changed the format., I personally don't mind and the other verbose outputs were to stdout so I left them the same. Normally I only separate them when there is a chance that the stdout messages are going to be piped to another process but not when they are unlikely to be piped.

drallensmith commented on Lib/test/test_zipfile.py:
```diff
@@ -2234,6 +2236,90 @@ def test_create_command(self):
+                out = self.zipfilecmd(opt, TESTFN2, pat.format(TESTFN))
+                #self.assertEqual(out, b'')
+                with zipfile.ZipFile(TESTFN2) as zf:
+                    self.assertEqual(len(zf.namelist()), expect)
```
Just checking len? And how about giving the namelist if the assert fails?

GadgetSteve replied:
There are other tests that are testing that files have the correct content so I felt that length was sufficient. Have given a more detailed assert message.

drallensmith commented on Lib/test/test_zipfile.py:
```diff
@@ -2234,6 +2236,90 @@ def test_create_command(self):
+                out = self.zipfilecmd('-v', opt, TESTFN2, pat.format(TESTFN))
+                #self.assertEqual(out, b'')
+                with zipfile.ZipFile(TESTFN2) as zf:
+                    self.assertGreaterEqual(len(zf.namelist()), expect,
```
Again, checking len only?

GadgetSteve replied:
So far, added some detail on assert.

drallensmith commented on Lib/test/test_zipfile.py:
```diff
@@ -2234,6 +2236,90 @@ def test_create_command(self):
+                #self.assertEqual(out, b'')
+                with zipfile.ZipFile(TESTFN2) as zf:
+                    self.assertGreaterEqual(len(zf.namelist()), expect,
+                                            "Testing Pattern %s" % pat.format(TESTFN))
```
A bit more informative message, such as actually giving the zf.namelist()?

GadgetSteve replied:
More detail added

drallensmith commented on Lib/test/test_zipfile.py:
```diff
@@ -2234,6 +2236,90 @@ def test_create_command(self):
+                with zipfile.ZipFile(TESTFN2) as zf:
+                    self.assertEqual(zf.namelist(), namelist)
+                    self.assertEqual(zf.read(namelist[0]), b'test 1')
+                    self.assertEqual(zf.read(namelist[2]), b'test 2')
```
For both of the above, how about giving the namelist[0]?

GadgetSteve replied:
More detail on assert added.

drallensmith commented on Lib/test/test_zipfile.py:
```diff
@@ -2234,6 +2236,90 @@ def test_create_command(self):
+                self.addCleanup(unlink, testfn3)
+                out = self.zipfilecmd('-v', opt, TESTFN2, testfn3)
+                with zipfile.ZipFile(TESTFN2) as zf:
+                    self.assertEqual(zf.read(testfn3), b"Spam Test!")
```
Give the filename so it can be checked if the assert fails (and I assume addCleanup won't remove it if there's an assert failure)?

GadgetSteve replied:
Done, not sure about clean-up.

drallensmith commented on Lib/zipfile.py:
```diff
@@ -1250,14 +1250,43 @@ def infolist(self):
+        """
+        Print a table of contents for the zip file.
+        :param file: Output file descriptor
+        :param verbose: How much detail to output"
```
`:param int verbose:`

GadgetSteve replied:
Done

drallensmith commented on Lib/zipfile.py:
```diff
@@ -1250,14 +1250,43 @@ def infolist(self):
+            ("CRC", 12, "{z.CRC:12}"),  # verbose = 3
+            ("Method", 11, "{compression:>11}"),  # verbose = 3
+        ]
+        elsement_dict = {0:1, 1:3, 2:4, 3:5, 4:6,}
```
"elsement"?

GadgetSteve replied:
Ooops.

drallensmith commented on Lib/zipfile.py:
```diff
@@ -1250,14 +1250,43 @@ def infolist(self):
+            ("Method", 11, "{compression:>11}"),  # verbose = 3
+        ]
+        elsement_dict = {0:1, 1:3, 2:4, 3:5, 4:6,}
+        fieldcount = elsement_dict.get(verbose, len(formatinfo))
```
So if verbose is out of the known range, print everything? You might make a comment to that effect.

GadgetSteve replied:
Added comment

drallensmith commented on Lib/zipfile.py:
```diff
@@ -1590,6 +1619,13 @@ def write(self, filename, arcname=None, compress_type=None):
+        if self.filename is not None:
+            if not isinstance(self.filename, int) and not isinstance(filename, int):
+                if os.path.abspath(filename) == os.path.abspath(self.filename):
+                    print("zipfile: Skipped adding %r as not good to zip self!" % filename)
```
"since it is the destination (recursive zips would be bad)"

GadgetSteve replied:
Clearer

drallensmith commented on Lib/zipfile.py:
```diff
@@ -1966,25 +2002,42 @@ def _compile(file, optimize=-1):
+        'If a directory is passed as a parameter to create or append',
+        'it will be recursively added. Wildcards are supported, glob style,',
+        'and a wildcard such as **/*.py will recurse.  All recursion will',
+        'skip any directories starting with . such as .git, etc.'
```
A flag to override this behavior would be nice (provided it still ignores os.curdir and os.pardir!). And you should probably add info on its behavior on encountering a symbolic link, if applicable on the given filesystem. (Note the possibility for recursion if it follows directory symbolic links!)

GadgetSteve replied:
Help message in both tarfile and zipfile updated to include information on how to override, `.**/*` and that this is as for glob.iglob() which is used.

drallensmith commented on Lib/zipfile.py:
```diff
@@ -2004,26 +2059,56 @@ def main(args=None):
-                for nm in os.listdir(path):
-                    addToZip(zf,
-                             os.path.join(path, nm), os.path.join(zippath, nm))
+                if not os.path.split(path)[-1].startswith('.'):
```
If verbose is set high enough, a warning about skipped directories (or included ones thanks to the flag I suggested above) would be good - that way, the user is more likely to realize when they should have included - or excluded - such directories.

GadgetSteve replied:
glob.iglob doesn't have an option to give me this information - I was trying to avoid using os.walk and walking, filtering myself. Added:
```
                        if file_name.startswith('**/'):
                            print("Recursing directories that don't start with .")
                        elif file_name.startswith('.**/'):
                            print("Recursing directories that do start with .")
```

drallensmith commented on Lib/zipfile.py:
```diff
@@ -2004,26 +2059,56 @@ def main(args=None):
+            else:
+                if args.verbose:
+                    print("Looking for wildcard matches to", path)
+                matches = glob.iglob(path)  # Might be a just wildcard
```
Might be just a wildcard?

GadgetSteve replied:
Comment made clearer and moved.

drallensmith commented on Lib/zipfile.py:
```diff
@@ -2004,26 +2059,56 @@ def main(args=None):
+                if args.verbose:
+                    print("Looking for wildcard matches to", path)
+                matches = glob.iglob(path)  # Might be a just wildcard
+                for nm in matches:  # if not then this will have no entries.
```
If something definitely is a wildcard, then not matching anything (either at all, or in a specifically named directory) should probably be reported.

GadgetSteve replied:
Done

drallensmith commented on Lib/zipfile.py:
```diff
@@ -2004,26 +2059,56 @@ def main(args=None):
+                    zippath = os.path.basename(path)
+                    if not zippath:
+                        zippath = os.path.basename(os.path.dirname(path))
+                    if zippath in ('', os.curdir, os.pardir):
```
Why check for '' and then replace it with itself?

GadgetSteve replied:
Done!

drallensmith reviewed (changes requested):
Thanks for the enhancements!

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
Changes in response to review comments from drallensmith

csabella committed:
bpo-31206: IDLE: Factor HighPage class from ConfigDialog (#3156)

Patch 2 of 3, to avoid horrendous diff.  Create highlights page from new HighPage class instead of old ConfigDialog methods and change tests to match.

csabella committed:
bpo-31206: IDLE: Factor HighPage class from ConfigDialog (#3160)

Part 3 of 3.  Remove old highlight functions and load_config as this functionality is now contained within classes. Patch by Cheryl Sabella.

orenmn committed:
bpo-28261: Fixed err msgs where PyArg_ParseTuple is used to parse normal tuples. (#3119)

segevfiner committed:
bpo-9566 & bpo-30747: Silence warnings from pyatomic.h macros (#3140)

* bpo-9566: Silence warnings from pyatomic.h macros

Apparently MSVC is too stupid to understand that the alternate branch is
not taken and emits a warning for it.

Warnings added in https://github.com/python/cpython/pull/2383

* bpo-9566: A better fix for the pyatomic.h warning

* bpo-9566: Remove a slash

Commit:
bpo-30923: Silence fall-through warnings included in -Wextra since gcc-7.0. (#3157)

vstinner committed:
bpo-31247: xmlrpc.server: break reference cycle (#3166)

xmlrpc.server now explicitly breaks reference cycles when using
sys.exc_info() in code handling exceptions.

orenmn committed:
bpo-31236: Improved some error messages of min() and max().

vstinner committed:
bpo-31249: test_concurrent_futures checks dangling threads (#3167)

Add a BaseTestCase class to test_concurrent_futures to check for
dangling threads and processes on all tests, not only tests using
ExecutorMixin.

vstinner committed:
bpo-31238: pydoc ServerThread.stop() now joins itself (#3151)

* bpo-31238: pydoc ServerThread.stop() now joins itself

ServerThread.stop() now joins itself to wait until
DocServer.serve_until_quit() completes and then explicitly sets
its docserver attribute to None to break a reference cycle.

* Add NEWS.d entry

vstinner committed:
Add test_subprocess.test_nonexisting_with_pipes() (#3133)

bpo-30121: Test the Popen failure when Popen was created with pipes.
Create also NONEXISTING_CMD variable in test_subprocess.py.

aragilar committed:
bpo-23835: Enforce that configparser defaults are strings (#2558)

* Enforce that configparser defaults are strings
* Update test_configparser.py

ambv committed:
bpo-23835: [docs] configparser converts defaults to strings (#3176)

Title says all.

ambv committed:
bpo-30983: [gdb] Fix py-bt, etc. for non-debug shared builds (#3153)

PEP 523 introduced _PyEval_EvalFrameDefault which inlines PyEval_EvalFrameEx on
non-debug shared builds.  This breaks the ability to use py-bt, py-up, and
a few other Python-specific gdb integrations.

This patch fixes the problem by only looking for _PyEval_EvalFrameDefault
frames.

test_gdb passes on both a debug and a non-debug build.

Original patch by Bruno "Polaco" Penteado.

vstinner committed:
bpo-31249: Fix ref cycle in ThreadPoolExecutor (#3178)

* bpo-31249: Fix ref cycle in ThreadPoolExecutor

concurrent.futures: WorkItem.run() used by ThreadPoolExecutor now
breaks a reference cycle between an exception object and the WorkItem
object. ThreadPoolExecutor.shutdown() now also clears its threads
set.

* shutdown() now only clears threads if wait is true.

* Revert changes on shutdown()

vstinner committed:
bpo-31234: test_httpservers joins the server thread (#3188)

vstinner committed:
bpo-31234: test_threaded_import: fix test_side_effect_import() (#3189)

* Don't leak the module into sys.modules
* Avoid dangling thread

brettcannon committed:
Touch up the contributing notes (#3158)

* Mention how to find out what to do
* Update URL
* Thanks contributors
* Explain there is no timeline

mjpieters committed:
bpo-31161: only check for parens error for SyntaxError (#3082)

Subclasses such as IndentError and TabError should not have this message
applied.

orenmn committed:
bpo-31229: Fixed wrong error messages when too many keyword arguments are received. (#3180)

methane committed:
bpo-31095: fix potential crash during GC (GH-2974)

ambv committed:
bpo-23835: Restore legacy defaults= behavior for RawConfigParser (#3191)

The fix for bpo-23835 fixed ConfigParser behavior in defaults= handling.
Unfortunately, it caused a backwards compatibility regression with
RawConfigParser objects which allow for non-string values.

This commit restores the legacy behavior for RawConfigParser only.

orenmn committed:
bpo-28261: fix err msgs where PyArg_ParseTuple is used to parse normal tuples (leftovers) (#3198)

orenmn committed:
bpo-29741: Update some methods in the _pyio module to also accept integer types. Patch by Oren Milman. (#560)

gpshead committed:
bpo-22536: Set the filename in FileNotFoundError. (#3194)

Have the subprocess module set the filename in the FileNotFoundError
exception raised on POSIX systems when the executable or cwd are missing.

gpshead committed:
Skip two tests not intended to pass on Windows. (#3202)

Commit:
bpo-30923: Silence fall-through warnings in libexpat build. (#3205)

Commit:
bpo-31275: Small refactoring to silence a fall-through warning. (#3206)

Commit:
bpo-31279: Silence -Wstringop-overflow warning. (#3207)

orenmn committed:
bpo-31271: Fix an assertion failure in io.TextIOWrapper.write. (#3201)

pfmoore committed:
bpo-31072: Rename the new filter argument for zipapp.create_archive. (#3049)

bpo-31072: Rename the new filter argument for zipapp.create_archive (GH-3049)

* Rename the new argument to "filter"
* Improve tests for the new functionality
* Add a "What's New" entry.

csabella committed:
bpo-30781: IDLE -  use ttk widgets in configdialog (#2654)

Patch by Cheryl Sabella.

terryjreedy committed:
bpo-31287: IDLE - do not alter tkinter.messagebox in configdialog tests. (#3220)

csabella committed:
bpo-30617: IDLE: docstrings and unittest for outwin.py (#2046)


Move some data and functions from the class to module level. Patch by Cheryl Sabella.

cryvate committed:
bpo-26656: Improve re.compile documentation (GH-3211)

- Link to the regular expressions object documentation
- Clarify that it can be used with more than the two methods currently stated.

pylessard committed:
bpo-30987 - Support for ISO-TP protocol in SocketCAN (#2956)

* Added support for CAN_ISOTP protocol

* Added unit tests for CAN ISOTP

* Updated documentation for ISO-TP protocol

* Removed trailing whitespace in documentation

* Added blurb NEWS.d file

* updated Misc/ACKS

* Fixed broken unit test that was using isotp const outside of skippable section

* Removed dependecy over third party project

* Added implementation for getsockname + unit tests

* Missing newline at end of ACKS file

* Accidentally inserted a type in ACKS file

* Followed tiran changes review #1 recommendations

* Added spaces after comma

pv committed:
bpo-10746: Fix ctypes PEP 3118 type codes for c_long, c_bool, c_int (#31)

Ctypes currently produces wrong pep3118 type codes for several types.
E.g. memoryview(ctypes.c_long()).format gives "<l" on 64-bit platforms,
but it should be "<q" instead for sizeof(c_long) == 8

The problem is that the '<>' endian specification in the struct syntax
also turns on the "standard size" mode, which makes type characters have
a platform-independent meaning, which does not match with the codes used
internally in ctypes.  The struct module format syntax also does not
allow specifying native-size non-native-endian items.

This commit adds a converter function that maps the internal ctypes
codes to appropriate struct module standard-size codes in the pep3118
format strings. The tests are modified to check for this.

gpshead committed:
Fix a typo in the Programming FAQ. (#3230)

subobjects, not subobjecs.

orenmn committed:
bpo-31243: Fixed PyArg_ParseTuple failure checks. (#3171)

serhiy-storchaka committed:
bpo-31286, bpo-30024: Fixed stack usage in absolute imports with (#3217)

binding a submodule to a name.

vanadium23 committed:
bpo-31065: Add doc about Popen.poll returning None. (#3169)

orenmn committed:
bpo-31291: Fixed an assertion failure in zipimport.zipimporter.get_data() (#3226)

if pathname.replace('/', '\\') returns non-string.

drallensmith committed:
bpo-5001: More-informative multiprocessing error messages (#3079)

* Make error message more informative

Replace assertions in error-reporting code with more-informative version that doesn't cause confusion over where and what the error is.

* Additional clarification + get travis to check

* Change from SystemError to TypeError

As suggested in PR comment by @pitrou, changing from SystemError; TypeError appears appropriate.

* NEWS file installation; ACKS addition (will do my best to justify it by additional work)

* Making current AssertionErrors in multiprocessing more informative

* Blurb added re multiprocessing managers.py, queues.py cleanup

* Further multiprocessing cleanup - went through pool.py

* Fix two asserts in multiprocessing/util.py

* Most asserts in multiprocessing more informative

* Didn't save right version

* Further work on multiprocessing error messages

* Correct typo

* Correct typo v2

* Blasted colon... serves me right for trying to work on two things at once

* Simplify NEWS entry

* Update 2017-08-18-17-16-38.bpo-5001.gwnthq.rst

* Update 2017-08-18-17-16-38.bpo-5001.gwnthq.rst

OK, never mind.

* Corrected (thanks to pitrou) error messages for notify

* Remove extraneous backslash in docstring.

terryjreedy committed:
bpo-31051:  Rearrange IDLE condigdialog GenPage into Window, Editor, and Help sections. (#3239)

crwilcox committed:
bpo-30581: Windows: os.cpu_count() returns wrong number of processors (#2934)

* Fixes #30581 by adding a path to use newer GetMaximumProcessorCount API on Windows calls to os.cpu_count()

* Add NEWS.d entry for bpo-30581, os.cpu_count on Windows.

* Tweak NEWS entry

alex committed:
Removed noop branch from ctypes code (#3234)

sir-sigurd committed:
Fix the indentation in Extending Python code example (GH-3244)

Code was indented with three spaces.
Fixed using four spaces.

pitrou committed:
[Trivial] Remove now redundant assert (#3245)

bugzPDX committed:
bpo-30737: Update DevGuide links to new URL (GH-3228)

Update old devguide links from https://docs.python.org/devguide to https://devguide.python.org

appeltel committed:
bpo-30096: Use ABC in abc reference examples (#1220)

Use base class rather than metaclass in examples.

vstinner committed:
bpo-31217: Fix regrtest -R for small integer (#3260)

Use a pool of integer objects toprevent false alarm when checking for
memory block leaks. Fill the pool with values in -1000..1000 which
are the most common (reference, memory block, file descriptor)
differences.

Co-Authored-By: Antoine Pitrou <pitrou@free.fr>

vstinner committed:
bpo-31250, test_asyncio: fix dangling threads (#3252)

* Explicitly call shutdown(wait=True) on executors to wait until all
  threads complete to prevent side effects between tests.
* Fix test_loop_self_reading_exception(): don't mock loop.close().
  Previously, the original close() method was called rather than the
  mock, because how set_event_loop() registered loop.close().

grzgrzgrz3 committed:
bpo-27144: concurrent.futures as_complete and map iterators do not keep reference to returned object (#1560)

* bpo-27144: concurrent.futures as_complie and map iterators do not keep
reference to returned object

* Some nits.  Improve wordings in docstrings and comments, and avoid relying on
sys.getrefcount() in tests.

vstinner committed:
bpo-31326: ProcessPoolExecutor waits for the call queue thread (#3265)

* bpo-31326: ProcessPoolExecutor waits for the call queue thread

concurrent.futures.ProcessPoolExecutor.shutdown() now explicitly
closes the call queue. Moreover, shutdown(wait=True) now also join
the call queue thread, to prevent leaking a dangling thread.

* Fix for shutdown() being called twice.

vstinner committed:
bpo-31250, test_asyncio: fix EventLoopTestsMixin.tearDown() (#3264)

Call doCleanups() to close the loop after calling
executor.shutdown(wait=True): see TestCase.set_event_loop() of
asyncio.test_utils.

Replace also gc.collect() with support.gc_collect().

vstinner committed:
bpo-31323: Fix reference leak in test_ssl (#3263)

Store exceptions as string rather than object to prevent reference
cycles which cause leaking dangling threads.

orenmn committed:
remove a redundant lower in urllib.parse.urlsplit (#3008)

serhiy-storchaka committed:
bpo-31185: Fixed miscellaneous errors in asyncio speedup module. (#3076)

pitrou committed:
Fix a c.f.as_completed() refleak previously introduced in bpo-27144 (#3270)

gpshead committed:
bpo-9146: add the missing NEWS entry. (#3275)

ned-deily committed:
bpo-12383: Also ignore __PYVENV_LAUNCHER__ (#3278)

Used in macOS framework builds.

segevfiner committed:
Add missing _sha3 module to Setup.dist (#2395)

benjaminp committed:
remove configure check for 'volatile' (#3281)

This is a required feature is C99, which we require.

benjaminp committed:
remove autoconf check for select() (#3283)

We never actually check HAVE_SELECT.

zmwangx committed:
bpo-31281: Fix pathlib.Path incompatibility in fileinput (gh-3208)

Fix fileinput with inplace=True to accept pathlib.Path objects.

alex committed:
Change code owners for hashlib and ssl to the crypto team (#3284)

* Change code owners for hashlib and ssl to the crypto team

* Include the core CSPRNG for the crypto-team

benjaminp committed:
remove check for bug last seem in Solaris 9 (#3285)

rhettinger committed:
Improve clarity (and small speed-up) by using tuple unpacking (#3289)

warsaw committed:
Use a team to maintain the email related packages. (#3290)

rhettinger committed:
Add references to modules I am responsible for (#3291)

bitdancer committed:
Clarify nature of parse_args 'args' argument. (#3292)

Patch by Paul.j3.  Includes an unrelated but useful addition to the
optparse porting section.

tiran committed:
bpo-25674: remove sha256.tbs-internet.com ssl test (#3297)

Signed-off-by: Christian Heimes <christian@python.org>

larryhastings committed:
Blurbify master branch. (#3298)

Blurbify master branch.

warsaw committed:
bpo-1198569: Allow string.Template braced pattern to be different (#3288)

* bpo-1198569: Allow the braced pattern to be different

``string.Template`` subclasses can optionally define ``braceidpattern`` if
they want to specify different placeholder patterns inside and outside the
braces.  If None (the default) it falls back to ``idpattern``.

benjaminp committed:
remote note about IRIX in aifc (#3299)

This comment hasn't been true since Python 3.0.

benjaminp committed:
remove configure.ac support for SGI_ABI (#3294)

zooba committed:
Fixes doc/make.bat to properly handle quoted paths. (#3302)

warsaw committed:
What's New for bpo-1198569 (#3303)

melvyn-sopacua committed:
bpo-30622: Change NPN detection: (#2079)

* Change NPN detection:

Version breakdown, support disabled (pre-patch/post-patch):
- pre-1.0.1: OPENSSL_NPN_NEGOTIATED will not be defined -> False/False
- 1.0.1 and 1.0.2: OPENSSL_NPN_NEGOTIATED will not be defined ->
False/False
- 1.1.0+: OPENSSL_NPN_NEGOTIATED will be defined and
OPENSSL_NO_NEXTPROTONEG will be defined -> True/False

Version breakdown support enabled (pre-patch/post-patch):
- pre-1.0.1: OPENSSL_NPN_NEGOTIATED will not be defined -> False/False
- 1.0.1 and 1.0.2: OPENSSL_NPN_NEGOTIATED will be defined and
OPENSSL_NO_NEXTPROTONEG will not be defined -> True/True
- 1.1.0+: OPENSSL_NPN_NEGOTIATED will be defined and
OPENSSL_NO_NEXTPROTONEG will not be defined -> True/True

* Refine NPN guard:

- If NPN is disabled, but ALPN is available we need our callback
- Make clinic's ssl behave the same way

This created a working ssl module for me, with NPN disabled and ALPN
enabled for OpenSSL 1.1.0f.

Concerns to address:
The initial commit for NPN support into OpenSSL [1], had the
OPENSSL_NPN_* variables defined inside the OPENSSL_NO_NEXTPROTONEG
guard. The question is if that ever made it into a release.
This would need an ugly hack, something like:

	#if defined(OPENSSL_NO_NEXTPROTONEG) && \
		!defined(OPENSSL_NPN_NEGOTIATED)
	#	define OPENSSL_NPN_UNSUPPORTED 0
	#	define OPENSSL_NPN_NEGOTIATED 1
	#	define OPENSSL_NPN_NO_OVERLAP 2
	#endif

[1] https://github.com/openssl/openssl/commit/68b33cc5c7

zware committed:
Cache externals, depending on changes to PCbuild (#3308)

serhiy-storchaka committed:
bpo-30502: Fix handling of long oids in ssl. (#2909)

zware committed:
Regen Moduls/clinic/_ssl.c.h (GH-3320)

Broken in GH-2079

rhettinger committed:
Code clean-up.  Remove unnecessary pre-increment before the loop starts. (#3312)

benjaminp committed:
remove IRIX support (closes bpo-31341) (#3310)

See PEP 11.

tiran committed:
Travis: use ccache (#3307)

rhettinger committed:
random_triangular:  sqrt() is more accurate than **0.5 (#3317)

ericsnowcurrently committed:
bpo-28411: Remove "modules" field from Py_InterpreterState. (#1638)

sys.modules is the one true source.

vstinner committed:
bpo-31170: Update libexpat from 2.2.3 to 2.2.4 (#3315)

* bpo-31170: Update libexpat from 2.2.3 to 2.2.4

Fix copying of partial characters for UTF-8 input (libexpat bug 115):
https://github.com/libexpat/libexpat/issues/115

* Add NEWS entry.

rhettinger committed:
Add comment to explain the implications of not sorting keywords (#3331)

In Python 3.6, sorted() was removed from _make_key() for the lru_cache and instead rely on guaranteed keyword argument order preservation.  This makes keyword argument handling faster but it also causes multiple callers with a different keyword argument order to be cached as separate items.  Depending on your point of view, this is either a performance regression (increased number of cache misses) or a performance enhancement (faster computation of keys).

rhettinger committed:
Fix terminology in comment and add more design rationale. (#3335)

* Fix terminology in comment and add more design rationale.

* Fix extra space

nascheme committed:
bpo-17852: Maintain a list of BufferedWriter objects.  Flush them on exit. (#1908)

* Maintain a list of BufferedWriter objects.  Flush them on exit.

In Python 3, the buffer and the underlying file object are separate
and so the order in which objects are finalized matters.  This is
unlike Python 2 where the file and buffer were a single object and
finalization was done for both at the same time.  In Python 3, if
the file is finalized and closed before the buffer then the data in
the buffer is lost.

This change adds a doubly linked list of open file buffers.  An atexit
hook ensures they are flushed before proceeding with interpreter
shutdown.  This is addition does not remove the need to properly close
files as there are other reasons why buffered data could get lost during
finalization.

Initial patch by Armin Rigo.

* Use weakref.WeakSet instead of WeakKeyDictionary.

* Simplify buffered double-linked list types.

* In _flush_all_writers(), suppress errors from flush().

* Remove NEWS entry, use blurb.

nascheme committed:
Revert "bpo-17852: Maintain a list of BufferedWriter objects.  Flush them on exit. (#1908)" (#3337)

This reverts commit e38d12ed34870c140016bef1e0ff10c8c3d3f213.

benjaminp committed:
bpo-31347: _PyObject_FastCall_Prepend: do not call memcpy if args might not be null (#3329)

Passing NULL as the second argument to to memcpy is undefined behavior even if the size is 0.

ned-deily committed:
Simplify NEWS entry to prevent suspicious warnings. (#3339)

ned-deily committed:
Include additional changes to support blurbified NEWS (#3340)

ned-deily committed:
Prevent a few make suspicious warnings. (#3341)

tiran committed:
bpo-30102: Call OPENSSL_add_all_algorithms_noconf (#3112)

The ssl and hashlib modules now call OPENSSL_add_all_algorithms_noconf() on
OpenSSL < 1.1.0. The function detects CPU features and enables optimizations
on some CPU architectures such as POWER8. Patch is based on research from
Gustavo Serra Scalet.

Signed-off-by: Christian Heimes <christian@python.org>

tiran committed:
bpo-31343: Include sys/sysmacros.h (#3318)

Include sys/sysmacros.h for major(), minor(), and makedev(). GNU C libray
plans to remove the functions from sys/types.h.

Signed-off-by: Christian Heimes <christian@python.org>

rhettinger committed:
Conceptually, roots is a set.  Also searching it as a set is a tiny bit faster (#3338)

gpshead committed:
Link to blurb on PyPI in the NEWS.d READMEs. (#3323)

vstinner committed:
_pickle: Fix whichmodule() (#3358)

_PyUnicode_FromId() can return NULL: replace Py_INCREF() with
Py_XINCREF().

Fix coverity report: CID 1417269.

jonathaneunice committed:
bpo-30662: fixed OrderedDict.__init__ docstring re PEP 468 (#2179)

* fixed OrderedDict.__init__ docstring re PEP 468

* tightened comment and mirrored to C impl

* added space after period per marco-buttu

* preserved substituted for stable

* drop references to Python 3.6 and PEP 468

benjaminp committed:
link to legacy doc on the non-legacy website (#3362)

tiran committed:
bpo-31320: No traceback to sys.stderr in test_ssl (#3360)

In case PROTOCOL_TLS_SERVER is used for both client context and server
context, the test thread dies with OSError. Catch OSError to avoid
traceback on sys.stderr

Signed-off-by: Christian Heimes <christian@python.org>

Commit:
bpo-31350: Optimize get_event_loop and _get_running_loop (#3347)

* call remove_done_callback in finally section

* Optimize get_event_loop and _get_running_loop

* rename _loop_pid as loop_pid and add blurb news

* rename _loop_pid as loop_pid and add blurb news

* add back _RunningLoop

* Update 2017-09-05-10-30-48.bpo-31350.dXJ-7N.rst

* Update 2017-09-05-10-30-48.bpo-31350.dXJ-7N.rst

vstinner committed:
bpo-31355: Travis CI: remove the macOS job (#3367)

ericsnowcurrently committed:
bpo-30860: Consolidate stateful runtime globals. (#2594)

* group the (stateful) runtime globals into various topical structs
* consolidate the topical structs under a single top-level _PyRuntimeState struct
* add a check-c-globals.py script that helps identify runtime globals

Other globals are excluded (see globals.txt and check-c-globals.py).

benjaminp committed:
correct initialization code (#3376)

Explicitly initialize struct members rather than relying on compiler extensions.

benjaminp committed:
pytime: include winsock2, so we can have a complete timeval type (#3377)

GadgetSteve committed:
Merge remote-tracking branch 'remotes/upstream/master' into fix-issue-300156

GadgetSteve committed:
Added options to select compression method on add or create

GadgetSteve committed:
Added options to select compression method on add or create

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

GadgetSteve committed:
Merge branch 'fix-issue-300156' of https://github.com/GadgetSteve/cpython into fix-issue-300156

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

merwok commented on Lib/tarfile.py:
```diff
@@ -1875,7 +1875,7 @@ def list(self, verbose=True, *, members=None):
         self._check()
 
-        if members is None:
+        if members is None or members == []:  # None or empty list
```
I think PEP 8 recommends to not add comments that just duplicate the code.  This one says what the code is doing, but not why it is desirable.

GadgetSteve replied:
The two sets of changes are largely duplicates and try to make the command lines equivalent between the two.
Good point on the comment, changed to why needed.
Spell Checked and a couple of minor changes to the News entry is that the doc that you had in mind?

merwok commented on Lib/tarfile.py:
```diff
@@ -2447,20 +2447,52 @@ def is_tarfile(name):
 def main():
     import argparse
+    compressions = {
+        # gz
```
Likewise, I don’t see the added values of the comments here.

GadgetSteve replied:
Removed

merwok commented on Lib/tarfile.py:
```diff
@@ -2447,20 +2447,52 @@ def is_tarfile(name):
 
     description = 'A simple command-line interface for tarfile module.'
-    parser = argparse.ArgumentParser(description=description)
+    epilog= ' '.join([
```
Tip: the Python compiler will automatically merge string literals that follow each other.

```python

epilog = ('Beginning of text, '
          'followed by more text.')
```

Here you also need dynamic text replacement; a simple way to do that is to use an f-string:

```python

available_compressions = ', '.join(compressions)
epilog = f'final extension (one of {available_compressions})'

GadgetSteve replied:
Good tips, (I guess I am still stuck in python 2 mode). Also reworked tarfile.main & zipfile.main docstrings to make more grammatical, (and better spelt). 

Also added note to zipfile epilogue that only a single compress can be used at invocation but later adds can use a different compression.

GadgetSteve replied:
Also only offer available compression methods in zipfile.main options.

GadgetSteve replied:
And only mention available compression methods as currently available.

merwok reviewed (commented):
Thanks for the patch!  Left some minor comments.  I think the doc would need a pass to improve the prose.

I wonder if it makes sense to add all these features in one PR.  Do they depend or build on each other, or was it simply convenient to work on them all at once?

I’m adding the maintainers of tarfile and zipfile as reviewers to get their opinio.

drallensmith commented:
I'll try to do a reread+edit of the docs, but am not sure when I can get around to it.

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
Changes following feedback from merwok plus only mention actually available compressions

drallensmith commented:
I may be able to work on documentation this weekend (family stuff prevented earlier). Which of the below need working on the most?

- Outputs during use of the command-line interface
- Clarification comments in the code
- Library documentation of the modules in general

serhiy-storchaka commented:
See my comments on the tracker. Discussions about the design should be on the tracker, GitHub is only for technical review of the concrete patch.

drallensmith commented:
@serhiy-storchaka - I'm not sure why documentation modifications wouldn't be part of the "concrete patch", but I'll take a look at the tracker when I have a moment.

brettcannon commented:
@drallensmith what Serhiy is saying is that asking what should go into the docs is a bugs.python.org topic to discussion. In other words if it isn't already in the pull request then it's discussion on this issue.

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156

GadgetSteve committed:
bpo-31182: Add --nozip64 flag to disable Zip64 processing.

merwok commented on Lib/zipfile.py:
```diff
@@ -2131,7 +2133,7 @@ def addToZip(zf, path, zippath):
         if len(files):
             count = 0
-            with ZipFile(zip_name, mode) as zf:
+            with ZipFile(zip_name, mode, allowZip64=args.zip64) as zf:
```
This works without a call to set_defaults?

GadgetSteve replied:
Yes @merwok - Because the argument parser is set to store_false on the flag ards.zip64 is automatically set to True unless the flag is seen.

GadgetSteve committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-300156
